### ISSUE TITLE
[Question] a line of code in loss.h

### ISSUE THREAD

**Issue by xiaolin-cheng** [davisking/dlib#1553]
Hi Davis,

To compute loss for loss_metric, why does the [iteration variable](https://github.com/davisking/dlib/blob/27a52936119f10d796c19b3d3df95b5a0e6040e7/dlib/dnn/loss.h#L1818) `c` start with `0`, not` r+1`? I thought it should be the same as the prior nested for loops.

Thanks.

**Comment by davisking** [comment]
You could write it that way if you changed the internals of the loop
appropriately.  I think this is a cleaner way to write the loop though.

**Comment by xiaolin-cheng** [comment]
Thanks @davisking ! 

When you calculate and sort distances of all pairs, the inner for loop [iteration variable](https://github.com/davisking/dlib/blob/27a52936119f10d796c19b3d3df95b5a0e6040e7/dlib/dnn/loss.h#L1776) starts from `r+1`, so the count of pairs is N*(N-1)/2. But when you calculate the loss, it starts from `0`, so the count of pairs is N*(N-1), double counted?

Thank you!

**Comment by davisking** [comment]
If you look closely you will see the gradients are all correct. It’s not doubling anything.

**Comment by xiaolin-cheng** [comment]
Thanks @davisking . A few questions:

1. How did you set the distance threshold as 0.6? Is it specific to the resnet34-like 128-element feature embedding? I am training resnet18 which has 512-element feature embedding. Would 0.6 still work? 

2. I would like port the training code to pytorch, but it doesn't seem to work. I am training on CASIA data and testing on LFW. Did you set learning rate 0.1 for SGD, and use learning rate decay?

Thank you very much!

**Comment by davisking** [comment]
I picked 0.6 empirically. Although the exact value doesn't really matter.

The exact code for training is discussed in the example programs for this
loss function.

**Comment by xiaolin-cheng** [comment]
Thanks @davisking . Did you pick margin 0.04 empirically too? Can I set it as 0? If I set threshold, say 0.5, can the training processing adapt to it automatically? 

Your mini-batch has N people and M images for each person, but you have an additional threshold (the largest positive distance). In my PyTorch training, I setup a mini-batch as balanced positive pairs plus negative pairs (roughly the same number of positive pairs and negative pairs). It doesn't seem to work.  Any significant problem here? 

Thank you very much! I really appreciate your great help!

**Comment by xiaolin-cheng** [comment]
@davisking BTW: any possibility to convert a dlib DNN model to PyTorch format? Thank you!

**Comment by davisking** [comment]
It's all empirical.  You can't set the margin to 0, that doesn't make sense. It must be some positive number.

I don't know why your pytorch code doesn't work.

You can dump dlib's models to xml and then convert them to any format you want from there.
